File: Java/14_Graph_DFS/LC_841_KeysAndRooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        int size = rooms.size();
        boolean[] visited = new boolean[size];
        dfs(0, rooms, visited);
        for (boolean flag : visited) {
            if (!flag) {
                return false;
            }
        }

        return true;
    }

    private void dfs(int key, List<List<Integer>> rooms, boolean[] visited) {
        if (visited[key]) {
            return;
        }

        visited[key] = true;

        for (int k : rooms.get(key)) {
            dfs(k, rooms, visited);
        }
    }
}

public class LC_841_KeysAndRooms {

}

File: Swift/Leetcode.playground/Sources/14_Graph_DFS/LC_841_KeysAndRooms.swift
class Solution {
    func canVisitAllRooms(_ rooms: [[Int]]) -> Bool {
        let size = rooms.count
        var visited = Array(repeating: false, count: size)
        dfs(0, rooms, &visited)
        for flag in visited {
            if !flag {
                return false
            }
        }
        return true
    }
    
    private func dfs(_ key: Int, _ rooms: [[Int]], _ visited: inout [Bool]) {
        if visited[key] {
            return
        }
        visited[key] = true
        for k in rooms[key] {
            dfs(k, rooms, &visited)
        }
    }
}